### ISSUE TITLE
Infrequent panic during long filesystem writes

### ISSUE THREAD

**Issue by sanderssj** [nanovms/nanos#1432]
During the write of a very large (>500MB) file, nanos will sometimes crash with a "Page fault on non-user memory".  I can reproduce this problem in master.
 
```
Page fault on non-user memory (vaddr 0xffff80002b000000)
cpu: 0
 interrupt: 000000000000000e (Page fault)
     frame: ffff800001720000
error code: 0000000000000000
   address: ffff80002b000000

   rax: 0000000000000000
   rbx: 0000000000000000
   rcx: ffff80002b000000
   rdx: 0000000000000200
   rsi: ffff80002b000000
   rdi: ffff80002a833000
   rbp: ffff80000171fd98
   rsp: ffff80000171fd70
    r8: 000000000000003f
    r9: 0000000000000000
   r10: 0000000000000040
   r11: 0000000000000000
   r12: ffff80002a833000
   r13: 0000000000000000
   r14: ffff800027dc1600
   r15: ffff80002b000000
   rip: ffffffff8002cc70        (runtime_memcpy + 00000000000003f0/000000000000062b)
rflags: 0000000000010046
    ss: 0000000000000000
    cs: 0000000000000008
    ds: 0000000000000000
    es: 0000000000000000
fsbase: 0000000000000000
gsbase: 0000000000000000

frame trace:
ffff80000171fda0:   ffffffff80030e6c    (sg_copy_to_buf + 000000000000007c/0000000000000152)
ffff80000171fde0:   ffffffff8001bce6    (pagecache_write_sg_finish + 0000000000000446/00000000000006c9)
ffff80000171ff40:   ffffffff8002d42c    (merge_join + 000000000000002c/0000000000000077)
ffff80000171ff60:   ffffffff8001c058    (pagecache_service_completions + 00000000000000e8/00000000000001c6)
ffff80000171ffa0:   ffffffff8001d7c2    (runloop_internal + 0000000000000152/0000000000000937)

stack trace:
ffff80000171fd70:   0000000000000200
ffff80000171fd80:   0000000000000200
ffff80000171fd90:   ffff80002b373dc0
ffff80000171fda0:   ffffffff80030e6c    (sg_copy_to_buf + 000000000000007c/0000000000000152)
ffff80000171fdb0:   ffff800027dc1700
ffff80000171fdc0:   0000000000000200
ffff80000171fdd0:   ffff800000e00000
ffff80000171fde0:   ffffffff8001bce6    (pagecache_write_sg_finish + 0000000000000446/00000000000006c9)
ffff80000171fdf0:   0000000000000200
ffff80000171fe00:   ffff80002b373dc0
ffff80000171fe10:   fffffffffffffe00
ffff80000171fe20:   0000000000000001
ffff80000171fe30:   0000000000000000
ffff80000171fe40:   ffff800029e8bc80
ffff80000171fe50:   0000000000000038
ffff80000171fe60:   000ffff80002a60e
```

**Comment by sanderssj** [comment]
While #1436 seems related to this, and appears to fix crashes with this frame trace, there is still an occasional panic with a page fault but a different frame trace.

**Comment by francescolavra** [comment]
@sanderssj did you see something similar to below? If so, the issue is fixed by #1476.
```
Page fault on non-user memory (vaddr 0xffffffffffffffff)
cpu: 0
 interrupt: 000000000000000e (Page fault)
     frame: ffff800000160000
error code: 0000000000000010
   address: ffffffffffffffff

   rax: ffffffffffffffff
   rbx: ffff8000356b1a00
   rcx: ffff8000249fffe0
   rdx: 0000000000000078
   rsi: ffff800000912500
   rdi: ffffffff800fffff	(bootstrap_region + 0000000000002f5f/0000000000200000)
   rbp: ffff80000015fdb8
   rsp: ffff80000015fd90
    r8: ffffffff800fece0	(bootstrap_region + 0000000000001c40/0000000000200000)
    r9: ffff800024a00000
   r10: 0000000000000040
   r11: 0000000000001de1
   r12: ffff800000912540
   r13: ffff800000912500
   r14: 0000000000000400
   r15: 00000000000006d8
   rip: ffffffffffffffff
rflags: 0000000000010092
    ss: 0000000000000000
    cs: 0000000000000008
    ds: 0000000000000000
    es: 0000000000000000
fsbase: 0000000000000000
gsbase: 0000000000000000

frame trace:
ffff80000015fdc0:   ffffffff8001d285	(pagecache_write_sg_finish + 0000000000000d95/000000000000100a)
ffff80000015ff50:   ffffffff80037248	(merge_join + 0000000000000038/0000000000000088)
ffff80000015ff70:   ffffffff80019742	(pagecache_service_completions + 00000000000000f2/00000000000001ea)
ffff80000015ffb0:   ffffffff800209e8	(runloop_internal + 0000000000000178/000000000000055b)

stack trace:
ffff80000015fd90:   ffffffff80052cf2	(flush_log_extension_complete + 0000000000000142/000000000000015e)
ffff80000015fda0:   ffff8000356b1a00
ffff80000015fdb0:   0000000000000069
ffff80000015fdc0:   ffffffff8001d285	(pagecache_write_sg_finish + 0000000000000d95/000000000000100a)
ffff80000015fdd0:   0000000000020a00
ffff80000015fde0:   0000000000000000
ffff80000015fdf0:   ffff800002a00000
ffff80000015fe00:   fffffffffffffe00
ffff80000015fe10:   ffff800000912580
ffff80000015fe20:   ffff800000e00088
ffff80000015fe30:   ffff80000000000c
ffff80000015fe40:   0000000000000000
ffff80000015fe50:   0000000000000069
ffff80000015fe60:   0000000000000000
ffff80000015fe70:   ffffffff800fd478	(bootstrap_region + 00000000000003d8/0000000000200000)
ffff80000015fe80:   0000000000000020
```

**Comment by francescolavra** [comment]
By the way, I occasionally see a similar panic also when running the fs_full runtime test.

**Comment by sanderssj** [comment]
Yes, the panics were very similar. I think that must be the problem. Thanks!

**Comment by francescolavra** [comment]
Closed by #1476
